fennel.lua

```diff
@@ -1069,6 +1069,7 @@ end
 local macroCurrentScope = GLOBAL_SCOPE
 
 local function macroexpand(ast, scope, once)
+    if not isList(ast) then return ast end -- bail early if not a list form
```

Comment:
Rather than bailing early, shouldn't we just skip the recursion at the end when we expand to a non-list? Bonus: avoiding early returns will make it easier when we port it to self-hosting.

Reply:
That occurred to me as well, but bailing early also makes it a bit more resilient. You can call `macroexpand` on an AST node without it blowing up in your face. If we want to give it stricter behavior, we could certainly move the `isList` check to the bottom. My update already skips recursion for a falsy AST node, so it would be a minor change, I just thought leaving the early return in makes this a bit safer to use.

Reply:
Hm; well if the `macroexpand` function can accept any AST form, not just a list, then why not call it from the top level of `compile1` instead of only inside the `isList` branch? Then the recursion would no longer be necessary, right?

Reply:
Ha! Fair point, I'm kinda double-dipping in the defensive coding and  `compile1` needs all the readability it can get. I'll reorganize things a bit.